helix-core/src/main/java/org/apache/helix/controller/stages/MessageGenerationPhase.java

```diff
@@ -335,15 +363,21 @@ private void addGeneratedMessageToMap(final Message message,
    */
   private void schedulePendingMessageCleanUp(
       final Map<String, Map<String, Message>> pendingMessagesToPurge, ExecutorService workerPool,
-      final HelixDataAccessor accessor) {
+      final HelixDataAccessor accessor, Map<String, Map<String, Message>> staleMessageMap) {
     workerPool.submit(new Callable<Object>() {
-      @Override public Object call() {
+      @Override
+      public Object call() {
         for (Map.Entry<String, Map<String, Message>> entry : pendingMessagesToPurge.entrySet()) {
           String instanceName = entry.getKey();
           for (Message msg : entry.getValue().values()) {
             if (accessor.removeProperty(msg.getKey(accessor.keyBuilder(), instanceName))) {
               LogUtil.logInfo(logger, _eventId, String
                   .format("Deleted message %s from instance %s", msg.getMsgId(), instanceName));
+              staleMessageMap.getOrDefault(msg.getTgtName(), Collections.emptyMap())
```

Comment:
Now I have a better understanding of your change. And I think this cleanup is not necessary.
The stale message map in the current state output object is re-calculated from the cache every time. So as long as we clean the cache properly, we don't need to care about the output map. Please correct me if I missed anything.

Reply:
yep, I agree. It'll go away later anyway. Shouldn't be a concern to leave it a little longer.